spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.nn
+
+import breeze.linalg.*
+import com.intel.analytics.bigdl.Module
+import com.intel.analytics.bigdl.nn.Graph.ModuleNode
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity, TensorModule}
+import com.intel.analytics.bigdl.tensor.{Storage, Tensor}
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.{T, Table}
+
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+/**
+ * Transformer model from "Attention Is All You Need".
+ * The Transformer model consists of an encoder and a decoder. Both are stacks
+ * of self-attention layers followed by feed-forward layers. This model yields
+ * good results on a number of problems, especially in NLP and machine translation.
+ * See "Attention Is All You Need" (https://arxiv.org/abs/1706.03762) for the full
+ * description of the model and the results obtained with its early version.
+ * @param hiddenSize
+ * @param numHeads
+ * @param filterSize
+ * @param numHiddenlayers
+ * @param postprocessDropout
+ * @param attentionDropout
+ * @param reluDropout
+ * @tparam T The numeric type in this module parameters.
+ */
+class TransformerLayer[T: ClassTag](
+   val vocabSize: Int,
+   val hiddenSize: Int,
+   val numHeads: Int,
+   val filterSize: Int,
+   val numHiddenlayers: Int,
+   val postprocessDropout: Float,
+   val attentionDropout: Float,
+   val reluDropout: Float,
+   val problem: TransformerType = LanguageModel)
+  (implicit ev: TensorNumeric[T]) extends BaseModule[T] {
+
+  override def buildModel(): Module[T] = {
+    problem match {
```

Comment:
type ?

Reply:
ok, change to "transformerType" as "type" is keyword.